include/grpc/event_engine/memory_allocator.h

```diff
@@ -24,24 +24,87 @@
 #include <vector>
 
 #include <grpc/event_engine/internal/memory_allocator_impl.h>
+#include <grpc/event_engine/slice.h>
 #include <grpc/slice.h>
-
-// forward-declaring an internal struct, not used publicly.
-struct grpc_slice_buffer;
+#include <grpc/slice_buffer.h>
 
 namespace grpc_event_engine {
 namespace experimental {
 
-// TODO(nnoble): needs implementation
+/// A Wrapper around \a grpc_slice_buffer pointer.
+///
+/// A slice buffer holds the memory for a collection of slices.
+/// The SliceBuffer object itself is meant to only hide the C-style API,
+/// and won't hold the data itself. In terms of lifespan, the
+/// grpc_slice_buffer ought to be kept somewhere inside the caller's objects,
+/// like a transport or an endpoint.
+///
+/// This lifespan rule is likely to change in the future, as we may
+/// collapse the grpc_slice_buffer structure straight into this class.
+///
+/// The SliceBuffer API is basically a replica of the grpc_slice_buffer's,
+/// and its documentation will move here once we remove the C structure,
+/// which should happen before the Event Engine's API is no longer
+/// an experimental API.
 class SliceBuffer {
  public:
-  SliceBuffer() { abort(); }
-  explicit SliceBuffer(grpc_slice_buffer*) { abort(); }
+  explicit SliceBuffer(grpc_slice_buffer* slice_buffer)
+      : slice_buffer_(slice_buffer) {}
+  SliceBuffer(const SliceBuffer& other) = delete;
+  SliceBuffer(SliceBuffer&& other) noexcept
+      : slice_buffer_(other.slice_buffer_) {
+    other.slice_buffer_ = nullptr;
+  }
+  /// Upon destruction, the underlying raw slice buffer is cleaned out and all
+  /// slices are unreffed.
+  ~SliceBuffer() { grpc_slice_buffer_destroy(slice_buffer_); }
+
+  /// Adds a new slice into the SliceBuffer and makes an attempt to merge
+  /// this slice with the last slice in the SliceBuffer.
+  void Add(Slice slice);
+
+  /// Adds a new slice into the SliceBuffer at the next available index.
+  /// Returns the index at which the new slice is added.
+  size_t AddIndexed(Slice slice);
+
+  /// Removes the first slice in the SliceBuffer.
+  void Pop() { grpc_slice_buffer_pop(slice_buffer_); }
+
+  /// Returns the number of slices held by the SliceBuffer.
+  size_t Count() { return slice_buffer_->count; }
+
+  /// Removes/deletes the last n bytes in the SliceBuffer.
+  void TrimEnd(size_t n) {
+    grpc_slice_buffer_trim_end(slice_buffer_, n, nullptr);
+  }
+
+  /// Move the first n bytes of the SliceBuffer into a memory pointed to by dst.
+  void MoveFirstIntoBuffer(size_t n, void* dst) {
+    grpc_slice_buffer_move_first_into_buffer(slice_buffer_, n, dst);
+  }
+
+  /// Removes and unrefs all slices in the SliceBuffer.
+  void Clear() { grpc_slice_buffer_reset_and_unref(slice_buffer_); }
+
+  /// Removes the first slice in the SliceBuffer and returns it.
+  Slice TakeFirst();
```

Comment:
How is this different to Pop?

Reply:
I removed Pop. Pop had removed the slice without returning it. This one removes and returns the first slice.